### PR TITLE
Ml on gcp 4 keras functional api

### PR DESCRIPTION
**Pull request by munnm**
finished work on Keras Functional API lab

### CODE REVIEW

File: courses/machine_learning/intro_to_tensorflow/4_keras_functional_api.ipynb [hunk 1]
```diff
@@ -0,0 +1,440 @@
+{
```
Comment (BenoitDherin) [inline courses/machine_learning/intro_to_tensorflow/4_keras_functional_api.ipynb]: <p>This is great! A couple of things:</p><p> </p><ul><li>Remove "Learn how to train a Keras model at scale on GCP" in the learning objectives</li><li>Remove "Once we have trained our model, we will see how to train our model at scale on GCP using AI Platform"</li><li>Add maybe a link to the original paper [Wide &amp; Deep Learning for Recommender Systems](<a href="https://arxiv.org/abs/1606.07792" target="_blank">https://arxiv.org/abs/1606.07792</a>)</li></ul><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/576/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (BenoitDherin) [inline courses/machine_learning/intro_to_tensorflow/4_keras_functional_api.ipynb]: <ul><li>Remove the following comments which don't add more to what the variable and function names are describing:</li><li class="ql-indent-1">&nbsp;&nbsp;# Compute Euclidean distance</li><li class="ql-indent-1">  # Add engineered features</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/576/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (BenoitDherin) [inline courses/machine_learning/intro_to_tensorflow/4_keras_functional_api.ipynb]: <p>I'd remove the positional argument "key=" (or "keys=") in the feature columns (it's always required in position 1) to make the code less verbose, e.g., </p><p> </p><p>fc.numeric_column(key="pickup_longitude")  --&gt; fc.numeric_column("pickup_longitude")</p><p> </p><p>fc.crossed_column(keys=[fc_bucketized_dlat, fc_bucketized_dlon], hash_bucket_size=NBUCKETS * NBUCKETS)</p><p> </p><p>--&gt; fc.crossed_column([fc_bucketized_dlat, fc_bucketized_dlon], hash_bucket_size=NBUCKETS * NBUCKETS)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/576/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

Comment (BenoitDherin) [inline courses/machine_learning/intro_to_tensorflow/4_keras_functional_api.ipynb]: <p>Same as above, I'd remove the "argument_name=" for all the positional arguments to avoid code verbosity.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/576/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (BenoitDherin) [inline courses/machine_learning/intro_to_tensorflow/4_keras_functional_api.ipynb]: <p>I'd import the need layers at the beginning to get rid of "tf.keras.layers" here in order to make the code super short and readable. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/576/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

Comment (BenoitDherin) [inline courses/machine_learning/intro_to_tensorflow/4_keras_functional_api.ipynb]: <p>I'd import the "Model" at the beginning to avoid "tf.keras". </p><p> </p><p>I'd package the model building into its own function that returns a compiled model: e.g.,</p><p> </p><p>def build_model(hidden_units):</p><p>  # etc.</p><p>  model = Model(inputs=etc.)</p><p>  model.compile(optimizer="adam", loss="mse", metrics=[rmse, "mse"])</p><p>  return model</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/576/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

Comment (BenoitDherin) [inline courses/machine_learning/intro_to_tensorflow/4_keras_functional_api.ipynb]: <p>Maybe have only one BATCH_SIZE parameter for both train and eval (it's set at 1000 in both cases here anyway) to decrease parameter complexity. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/576/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

Comment (BenoitDherin) [inline courses/machine_learning/intro_to_tensorflow/4_keras_functional_api.ipynb]: <p>I'd put the model compilation into the build_model function referenced above.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/576/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

Comment (BenoitDherin) [inline courses/machine_learning/intro_to_tensorflow/4_keras_functional_api.ipynb]: <p>I don't like that this code depends on the number of examples in the dataset in a hardcoded way. We need to change that variable manually if we switch the dataset (e.g. a smaller dataset may crash the training, which happened to me in the training at scale with the same code pattern). Also, we need to know the exact number of examples ahead of time, and pass it to the code, which adds complexity when we package the code into a container later on)</p><p>Not sure the best way to address that though... any idea?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/576/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='20'/>

Comment (lakshmanok) [reply courses/machine_learning/intro_to_tensorflow/4_keras_functional_api.ipynb]: <p>This is not size of training set. It is total number of training examples to train on. I'll write an article explaining difference and why we do it the way. The crash is a bug that we should fix. The concept is correct, though. </p>

Comment (munnm) [reply courses/machine_learning/intro_to_tensorflow/4_keras_functional_api.ipynb]: these all look great. Thanks Benoit!!
